### PR TITLE
[4.0]  Add prepared statements for mod tags popular

### PR DESCRIPTION
**Pull request by HLeithner**
### Summary of Changes
Updated SQL queries to prepared statements and made some cleanups around the queries.

### Testing Instructions
Use the module in all ways you can think of.

* create a frontend module
* create articles with tags
* check the tags in the module if its the same without the PR

### Expected result
Nothing changed.

### CODE REVIEW

File: modules/mod_tags_popular/Helper/TagsPopularHelper.php [hunk 1]
```diff
@@ -35,26 +36,26 @@ public static function getList(&$params)
...
 				array(
 					'MAX(' . $db->quoteName('tag_id') . ') AS tag_id',
-					' COUNT(*) AS count', 'MAX(t.title) AS title',
+					' COUNT(*) AS count', 'MAX(' . $db->quoteName('t.title') . ') AS title',
 					'MAX(' . $db->quoteName('t.access') . ') AS access',
 					'MAX(' . $db->quoteName('t.alias') . ') AS alias',
 					'MAX(' . $db->quoteName('t.params') . ') AS params',
 				)
 			)
 			->group($db->quoteName(array('tag_id', 'title', 'access', 'alias')))
```
Comment (Quy) [inline modules/mod_tags_popular/Helper/TagsPopularHelper.php]:
```suggestion
			->group($db->quoteName(['tag_id', 'title', 'access', 'alias']))
```

File: modules/mod_tags_popular/Helper/TagsPopularHelper.php [hunk 2]
```diff
@@ -116,22 +123,24 @@ public static function getList(&$params)
...
 				$equery = $db->getQuery(true)
 					->select(
-						array(
-							'a.tag_id',
-							'a.count',
-							'a.title',
-							'a.access',
-							'a.alias',
+						$db->quoteName(
+								[
```
Comment (Quy) [inline modules/mod_tags_popular/Helper/TagsPopularHelper.php]:
```suggestion
							[
```

File: modules/mod_tags_popular/Helper/TagsPopularHelper.php [hunk 3]
```diff
@@ -77,15 +80,16 @@ public static function getList(&$params)
...
 
 		if ($timeframe !== 'alltime')
 		{
-			$query->where($db->quoteName('tag_date') . ' > ' . $query->dateAdd($db->quote($nowDate), '-1', strtoupper($timeframe)));
+			$query->where($db->quoteName('tag_date') . ' > ' . $query->dateAdd(':nowDate1', '-1', strtoupper($timeframe)))
+				->bind(':nowDate1', $nowDate);
 		}
 
-		$query->join('INNER', $db->quoteName('#__tags', 't') . ' ON ' . $db->quoteName('tag_id') . ' = t.id')
+		$query->join('INNER', $db->quoteName('#__tags', 't') . ' ON ' . $db->quoteName('tag_id') . ' = ' . $db->quoteName('t.id'))
```
Comment (Quy) [inline modules/mod_tags_popular/Helper/TagsPopularHelper.php]: Use `innerJoin`

File: modules/mod_tags_popular/Helper/TagsPopularHelper.php [hunk 4]
```diff
@@ -116,22 +123,24 @@ public static function getList(&$params)
 				$query->order('count DESC');
```
Comment (Quy) [inline modules/mod_tags_popular/Helper/TagsPopularHelper.php]: Apply `$db->quoteName`?

**Review by alikon — APPROVED** [review]

**Comment by alikon** [comment]
I have tested this item :white_check_mark: successfully on 86ee1698f83b378bf47b69066e26c2f020ada2e1<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25043">issues.joomla.org/tracker/joomla-cms/25043</a>.</sub>

**Comment by Quy** [comment]
> Error: Unknown column 'MAX(`tag_id`) AS tag_id' in 'field list': Unknown column 'MAX(`tag_id`) AS tag_id' in 'field list'

**Comment by Quy** [comment]
> Warning: Illegal offset type in \libraries\vendor\joomla\database\src\Mysqli\MysqliQuery.php on line 123

> Warning: mysqli_stmt::bind_param(): Number of variables doesn't match number of parameters in prepared statement in \libraries\vendor\joomla\database\src\Mysqli\MysqliStatement.php on line 425

**Comment by alikon** [comment]
could be related to this line https://github.com/joomla/joomla-cms/pull/25043/files#diff-b76ce82322bfdac12f53dcffdcc8a16cR104  the actual `bind()`  doesn't allow  what has been made here
probably a composer update will solve

**Comment by HLeithner** [comment]
I closed my own PR https://github.com/joomla-framework/database/pull/162 accidentally, this one is needed for this PR to work. 

Applying the database Pr makes this one work again...

**Comment by alikon** [comment]
When I've tested this one I should have the framework patched


Il ven 21 giu 2019, 10:50 Harald Leithner <notifications@github.com> ha
scritto:

File: modules/mod_tags_popular/Helper/TagsPopularHelper.php [hunk 5]
```diff
@@ -35,36 +36,36 @@ public static function getList(&$params)
...
-				array(
-					'MAX(' . $db->quoteName('tag_id') . ') AS tag_id',
-					' COUNT(*) AS count', 'MAX(t.title) AS title',
-					'MAX(' . $db->quoteName('t.access') . ') AS access',
-					'MAX(' . $db->quoteName('t.alias') . ') AS alias',
-					'MAX(' . $db->quoteName('t.params') . ') AS params',
-				)
+				[
+					'MAX(' . $db->quoteName('tag_id') . ') AS ' . $db->quoteName('tag_id'),
+					' COUNT(*) AS count', 'MAX(' . $db->quoteName('t.title') . ') AS ' . $db->quoteName('title'),
```
Comment (Quy) [inline modules/mod_tags_popular/Helper/TagsPopularHelper.php]: Is it necessary to apply `quoteName` to the `AS` alias value? 
`count` does not have `quoteName`. 
Remove space before `COUNT`.

File: modules/mod_tags_popular/Helper/TagsPopularHelper.php [hunk 6]
```diff
@@ -77,15 +78,16 @@ public static function getList(&$params)
...
 
 		if ($timeframe !== 'alltime')
 		{
-			$query->where($db->quoteName('tag_date') . ' > ' . $query->dateAdd($db->quote($nowDate), '-1', strtoupper($timeframe)));
+			$query->where($db->quoteName('tag_date') . ' > ' . $query->dateAdd(':nowDate1', '-1', strtoupper($timeframe)))
+				->bind(':nowDate1', $nowDate);
 		}
 
-		$query->join('INNER', $db->quoteName('#__tags', 't') . ' ON ' . $db->quoteName('tag_id') . ' = t.id')
+		$query->join('INNER', $db->quoteName('#__tags', 't') . ' ON ' . $db->quoteName('tag_id') . ' = ' . $db->quoteName('t.id'))
```
Comment (Quy) [inline modules/mod_tags_popular/Helper/TagsPopularHelper.php]:
```suggestion
		$query->innerJoin($db->quoteName('#__tags', 't'), $db->quoteName('tag_id') . ' = ' . $db->quoteName('t.id'))
```

File: modules/mod_tags_popular/Helper/TagsPopularHelper.php (hunk 5 above)
Comment (HLeithner) [reply modules/mod_tags_popular/Helper/TagsPopularHelper.php]: it's a name property so I think so but I'm not sure if it's needed. Most of the parts in J use it so I do it too.

File: modules/mod_tags_popular/Helper/TagsPopularHelper.php [hunk 7]
```diff
@@ -35,36 +36,37 @@ public static function getList(&$params)
...
-				array(
-					'MAX(' . $db->quoteName('tag_id') . ') AS tag_id',
-					' COUNT(*) AS count', 'MAX(t.title) AS title',
-					'MAX(' . $db->quoteName('t.access') . ') AS access',
-					'MAX(' . $db->quoteName('t.alias') . ') AS alias',
-					'MAX(' . $db->quoteName('t.params') . ') AS params',
-				)
+				[
+					'MAX(' . $db->quoteName('tag_id') . ') AS ' . $db->quoteName('tag_id'),
+					'COUNT(*) ' . $db->quoteName('count'),
```
Comment (SharkyKZ) [inline modules/mod_tags_popular/Helper/TagsPopularHelper.php]:
```suggestion
					'COUNT(*) AS ' . $db->quoteName('count'),
```

File: modules/mod_tags_popular/Helper/TagsPopularHelper.php [hunk 8]
```diff
@@ -77,30 +79,32 @@ public static function getList(&$params)
...
-			$query->where($db->quoteName('tag_date') . ' > ' . $query->dateAdd($db->quote($nowDate), '-1', strtoupper($timeframe)));
+			$query->where($db->quoteName('tag_date') . ' > ' . $query->dateAdd(':nowDate1', '-1', strtoupper($timeframe)))
+				->bind(':nowDate1', $nowDate);
 		}
 
-		$query->join('INNER', $db->quoteName('#__tags', 't') . ' ON ' . $db->quoteName('tag_id') . ' = t.id')
-			->join(
-				'INNER',
-				$db->quoteName('#__ucm_content', 'c') . ' ON ' . $db->quoteName('m.core_content_id') . ' = ' . $db->quoteName('c.core_content_id')
+		$query->innerJoin($db->quoteName('#__tags', 't'), $db->quoteName('tag_id') . ' = ' . $db->quoteName('t.id'))
```
Comment (Quy) [inline modules/mod_tags_popular/Helper/TagsPopularHelper.php]: Change innerJoin to join. Repeat below.

File: modules/mod_tags_popular/Helper/TagsPopularHelper.php [hunk 9]
```diff
@@ -77,30 +79,32 @@ public static function getList(&$params)
...
-			->where('(' . $db->quoteName('c.core_publish_up') . ' = ' . $nullDate
-				. ' OR ' . $db->quoteName('c.core_publish_up') . ' <= ' . $db->quote($nowDate) . ')'
+			->where('(' . $db->quoteName('c.core_publish_up') . ' = :nullDate2'
+				. ' OR ' . $db->quoteName('c.core_publish_up') . ' <= :nowDate2)'
 			)
-			->where('(' . $db->quoteName('c.core_publish_down') . ' = ' . $nullDate
-				. ' OR  ' . $db->quoteName('c.core_publish_down') . ' >= ' . $db->quote($nowDate) . ')'
-			);
+			->where('(' . $db->quoteName('c.core_publish_down') . ' = :nullDate3'
+				. ' OR  ' . $db->quoteName('c.core_publish_down') . ' >= :nowDate3)'
```
Comment (Quy) [inline modules/mod_tags_popular/Helper/TagsPopularHelper.php]:
```suggestion
				. ' OR ' . $db->quoteName('c.core_publish_down') . ' >= :nowDate3)'
```

File: modules/mod_tags_popular/Helper/TagsPopularHelper.php [hunk 10]
```diff
@@ -118,22 +121,24 @@ public static function getList(&$params)
 				$query->order('count DESC');
```
Comment (SharkyKZ) [inline modules/mod_tags_popular/Helper/TagsPopularHelper.php]:
```suggestion
				$query->order($db->quoteName('count') . ' DESC');
```

**Comment by SharkyKZ** [comment]
I have tested this item :white_check_mark: successfully on 19d980f089a44365b2071e86b7e4fb2e27ce18c3<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25043">issues.joomla.org/tracker/joomla-cms/25043</a>.</sub>

**Comment by ghost** [comment]
@alikon can you please retest?

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 19d980f089a44365b2071e86b7e4fb2e27ce18c3<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25043">issues.joomla.org/tracker/joomla-cms/25043</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25043">issues.joomla.org/tracker/joomla-cms/25043</a>.</sub>

**Comment by wilsonge** [comment]
Thanks!

**Pull request merged**